src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2523,6 +2521,59 @@ private bool Fix(int iParam, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
             }
         }
 
+        /// <summary>
+        /// Takes the dynamic flags from both types and applies them onto the target.
+        /// </summary>
+        internal static TypeSymbol MergeDynamic(TypeSymbol first, TypeSymbol second, TypeSymbol target, Compilation compilation)
+        {
+            // SPEC: 4.7 The Dynamic Type
+            //       Type inference (7.5.2) will prefer dynamic over object if both are candidates.
+            if (first.Equals(second, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreDynamic))
+            {
+                return target;
+            }
+            ImmutableArray<bool> flags1 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(first, RefKind.None);
+            ImmutableArray<bool> flags2 = CSharpCompilation.DynamicTransformsEncoder.EncodeWithoutCustomModifierFlags(second, RefKind.None);
+            Debug.Assert(flags1.Length == flags2.Length);
+            ImmutableArray<bool> mergedFlags = flags1.SelectAsArray((f, index, other) => f | other[index], flags2);
+
+            return DynamicTypeDecoder.TransformTypeWithoutCustomModifierFlags(target, (AssemblySymbol)compilation.Assembly, RefKind.None, mergedFlags);
```

Comment:
Does `DynamicTypeDecoder` and `TupleTypeDecoder` require the source assembly or the corlib assembly? (It looks like all uses are for `GetSpecialType` which use corlib. In that case, `ConversionsBase` already has a `corLibrary` field that could be used by callers.